src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -315,7 +384,9 @@ public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListene
                     Run<?,?> r = it.next();
                     ok |= perform(r, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console);
                 }
-                return ok;
+                if (!ok) {
+                    throw new CopyArtifactException(Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter));
```

Comment:
Can be inlined into the loop. I see no reason to continue copying if one iteration fails.